sw/device/lib/dif/dif_keymgr.h

```diff
@@ -0,0 +1,683 @@
+ * `dif_keymgr_advance_state()` before keys or identity seeds can be generated.
+ *
+ * This is an asynchronous function because key manager state transitions
+ * involve KMAC operations that can take some time to complete. Clients must
+ * check the status of key manager using `dif_keymgr_get_status()` before
+ * calling other functions in this library.
+ *
+ * The entropy source must be initialized before this function is called because
+ * key manager relies on the entropy source to initialize its secret value.
+ *
+ * This function must be called only once per power cycle.
+ *
+ * @param keymgr A key manager handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_result_t dif_keymgr_initialize_state(const dif_keymgr_t *keymgr);
+
+/**
+ * Valid lengths for the `binding_value` array in `dif_keymgr_state_params_t`.
+ */
+typedef enum dif_keymgr_binding_value_len {
+  /**
+   * `binding_value` length when transitioning to `OwnerIntermediateKey` and
+   * `OwnerRootKey` states.
+   */
+  kDifKeymgrBindingValueLenTag = 4,
+  /**
+   * `binding_value` length when transitioning to `CreatorRootKey` state.
+   */
+  kDifKeymgrBindingValueLenRomExtHash = 8,
+} dif_keymgr_binding_value_len_t;
+
+/**
+ * Parameters for a key manager state.
+ */
+typedef struct dir_keymgr_state_params {
+  /**
+   * Length of the `binding_value` array.
+   */
+  size_t binding_value_len;
+
+  /**
+   * This value is used by key manager to derive secret values and can be either
+   * a value that represents the contents of a boot stage, e.g. a (truncated)
+   * hash, or a tag.
+   *
+   * If it is a hash, changes in a boot stage will change the secret value, and
+   * consequently the versioned keys and identity seeds generated at subsequent
+   * boot stages. If it is a tag, those secret values, versioned keys, and
+   * identity seeds will be preserved across updates of the boot stage as long
+   * as the tag remains the same.
+   *
+   * For transitioning to CreatorRootKey state, the length of this array must be
+   * `kDifKeymgrBindingValueLenRomExtHash`. For all other states, its length
+   * must be `kDifKeymgrBindingValueLenTag`.
+   */
+  const uint32_t *binding_value;
+
+  /**
+   * Maximum allowed version for keys generated at the next state.
+   */
+  uint32_t max_key_version;
+} dif_keymgr_state_params_t;
+
+/**
+ * Sets parameters for a key manager state.
+ *
+ * This function sets and locks the max version value and the binding value that
+ * is used to compute the secret value for the operational states of key
+ * manager, i.e. CreatorRootKey, OwnerIntermediateKey, and OwnerRootKey.
+ *
+ * @param keymgr A key manager handle.
+ * @param state A key manager state.
+ * @param params State parameters.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_lockable_result_t dif_keymgr_set_state_params(
+    const dif_keymgr_t *keymgr, dif_keymgr_state_t state,
+    dif_keymgr_state_params_t params);
```

Comment:
I initially had this as a part of `dif_keymgr_advance_state()` but based on our discussions w/ @tjaychen and @moidx we will probably need a way to set and lock state inputs before advancing state due to lack of entropy. `dif_keymgr_state_params_t` does not hold the parameters of every state because every stage in the secure boot chain will set the parameters of the next state.

Regarding locking: There is no need to postpone locking but we may want to handle it separately by defining `*_lock` and `*_is_locked` depending on how we want to handle hardening.

Reply:
I decided to remove this function because there is a single register for the binding value now.